pkgs/shells/nushell/plugins/default.nix

```diff
@@ -22,25 +24,154 @@ lib.makeScope newScope (
+                    local old="''${1${''#''}v}" new="''${2${''#''}v}"
+
+                    local oM oN oP nM nN nP
+                    IFS=. read -r oM oN oP <<<"$old"
+                    IFS=. read -r nM nN nP <<<"$new"
+
+                    local cmp
+                    if   (( nM > oM )); then cmp=1
+                    elif (( nM < oM )); then cmp=-1
+                    elif (( nN > oN )); then cmp=1
+                    elif (( nN < oN )); then cmp=-1
+                    elif (( nP > oP )); then cmp=1
+                    elif (( nP < oP )); then cmp=-1
+                    else cmp=0
+                    fi
+
+                    if (( cmp < 0 )); then
+                      return 1
+                    fi
+
+                    if (( cmp == 0 )); then
+                      return 0
+                    fi
+
+                    if (( oM == 0 && nM == 0 )); then
+                      if (( nN != oN )); then
+                        return 1
+                      else
+                        return 0
+                      fi
+                    fi
+
+                    if (( oM == 0 && nM >= 1 )); then
+                      return 1
+                    fi
+                    if (( nM != oM )); then
+                      return 1
+                    fi
+                    return 0
+                  }
+
+                  is_compatible() {
+                    local name="$1" want="$2"
+                    [ -f "$root_dir/Cargo.lock" ] || return 1
+                    local locked
+                    locked="$(
+                      awk -v n="$name" '
+                        /^\[\[package\]\]/{inpkg=1; nm=""; vr=""}
+                        inpkg && /^name = "/    { nm=$0; sub(/^name = "/,"",nm); sub(/"$/,"",nm) }
+                        inpkg && /^version = "/ { vr=$0; sub(/^version = "/,"",vr); sub(/"$/,"",vr) }
+                        inpkg && NF==0          { if(nm==n){print vr; exit 0} inpkg=0 }
+                        END { if(nm==n) print vr }
+                      ' "$root_dir/Cargo.lock"
+                    )" || return 1
+                    [ -z "$locked" ] && return 1
+
+                    is_semvers_compatible "$locked" "$want"
+                    case $? in
+                      0) return 0 ;;
+                      1) return 1 ;;
+                      *) return 0 ;;
+                    esac
+                  }
+
+                  patch_crate_version() { # crate, version
+                    local crate="$1" ver="$2"
+
+                    find "$root_dir" -name Cargo.toml -print0 |
+                    while IFS= read -r -d ''' toml; do
+                      sed -i -E "s|^([[:space:]]*$crate[[:space:]]*=[[:space:]]*\")[^\"]*(\".*)|\1$ver\2|" "$toml" || true
+                      sed -i -E "s|^([[:space:]]*$crate[[:space:]]*=[[:space:]]*\{[^}]*version[[:space:]]*=[[:space:]]*\")[^\"]*(\"[^}]*\})|\1$ver\2|" "$toml" || true
+                    done
+                  }
```

Comment:
You should probably use [`cargo-edit`](https://github.com/killercup/cargo-edit/) for this, something like:

```ShellSession
$ cargo upgrade -p "$pkg@$ver"
```